docs/designs/local-build.md

```diff
@@ -0,0 +1,142 @@
+files: '**/*.md'
+xref:
+- https://api.docs.com/xref/dotnet
+- https://api.docs.com/xref/java
+```
+
+> For backward compatibility with v2 configs, we automatically add `use: https://api.docs.com/config` whenever `.openpublishing.config.json` file exists
+
+The response is a JSON object that confirms to `docfx.yml` configuration schema.
+
+_Example Request:_
+
+```
+GET /config/docfx
+```
+
+_Example Response:_
+```json
+{
+   "monikerDefinition": "https://api.docs.com/config/monikerDefinition",
+   "metadataSchema": "https://api.docs.com/config/metadataSchema",
+   "customErrors": {
+     "h1-missing": {
+       "severity": "error"
+     }
+   }
+}
+```
+
+##### Config parameterization
+
+Server side config is different between repos, docsets or locals. `docfx` parameterizes outgoing HTTP requests using HTTP headers:
+
+name | example
+----|----
+`Docfx-Name` | Docs.azure-documents
+`Docfx-Locale` | en-us
+`Docfx-Repository-Url` | https://github.com/MicrosoftDocs/azure-docs
+`Docfx-Repository-Branch` | master
+
+> An alternative is to pass these parameters explicitly use URL parameters, such design may produce tedious config like `https://api.docs.com/config/docfx?name={name}&locale={locale}&repository_url={repository_url}&repository_branch={repository_branch}` and it is hard to version.
+
+> The practise of using `X-` as custom http headers is __deprecated__, thus our headers are prefixed by `Docfx_`.
+
+##### Config Authentication
+
+`docfx` access protected resources using standard HTTP headers as described in [Credential.md](./credential.md). The config service exposed by build service uses the current OPS token based authentication.
+
+## vscode Extension
+
+The docfx vscode extension provides rich authoring tools when working with docs repos inside vscode. The initial phase is to provide a simple command that can trigger a build and show diagnostics in vscode problems window. As we progress, we can light up more advanced features like providing diagnostics as typing, providing intellisense, etc.
+
+### Acquisition
+
+The vscode extension is initially a standalone extension, but will be bundled into `docs-authoring-pack` after getting matrured.
+
+Users don't need to install anything particular after installing the vscode extension. This is done by publishing docfx as a self contained, platform specific executable to a well known blob storage location. The vscode extension downloads and unzips latest docfx for the current platform on `extension activation`. This allows the vscode extension to ship independently from docfx and ensure the latest version of docfx is used.
+
+### Trigger Build
+
+Handling build command is divided into two phases given the time constraint.
+
+#### Phase 1
+
+For the initial phase, the vscode extension calls `docfx build` and `docfx restore` directly, parses generated report files and presents them to problems window. It pipelines `docfx` console output to a specific output channel.
+
+#### Phase 2
```

Comment:
>Phase 2 [](start = 5, length = 7)

Here is the VS Code description about Language server:
> Language Server is a special kind of Visual Studio Code extension that powers the editing experience for many programming languages. With Language Servers, you can implement autocomplete, error-checking (diagnostics), jump-to-definition, and many other [language features](https://code.visualstudio.com/api/language-extensions/programmatic-language-features) supported in VS Code

Seem the `build`, `authentication`, `restore` is not a kind of Editing experience of a programming language, and vscode-language-client doesn't provide native support for those features.

on another hand, handling those request may block the language server from responding the real editing experience request.

maybe we need to reconsider whether it is necessary to consolidate all the feature onto the same pipeline using language server.

Reply:
There are two layers, the __transport layer__ (JSON RPC) and the __protocol layer__ (LSP on top of JSON PRC). We are building interactions between the server and the client using the same __transport layer__, leveraging the _LSP protocol_ as much as we can, building custom protocols as needed.

Reply:
so we are in the same line about the `build`/`restore`/'auth' is not a language server provided feature, right?
talk about leveraging the LSP to achieve those scenarios in the docfx serve mode, how can we resolve the second problem?

---
In reply to: [343428643](https://github.com/dotnet/docfx/pull/5137#discussion_r343428643) [](ancestors = 343428643)

Reply:
For the blocking problem,  `JSON RPC` can handle concurrent requests/notifications, one API does not block other APIs. Also, long operations can be modeled as async API, e.g.,`build` triggers a build and immediately returned with a `build_id`, then you can correlate progress and diagnostics associated with that `build_id`.